src/audio/selector/selector.c

```diff
@@ -692,9 +699,84 @@ static void set_selector_params(struct comp_dev *dev,
 	buffer_release(source);
 }
 
+static int selector_verify_params(struct processing_module *mod,
+				  struct sof_ipc_stream_params *params)
+{
+	struct comp_dev *dev = mod->dev;
+	struct comp_data *cd = module_get_private_data(mod);
+	struct comp_buffer *buffer, *sinkb;
+	struct comp_buffer __sparse_cache *buffer_c, *sink_c;
+	uint32_t in_channels;
+	uint32_t out_channels;
+
+	comp_dbg(dev, "selector_verify_params()");
+
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
+				source_list);
+
+	/* Apply input/output channels count */
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK) {
+		/* fetch sink buffer for playback */
+		buffer = list_first_item(&dev->bsink_list, struct comp_buffer,
+					 source_list);
+		in_channels = cd->config.in_channels_count;
+
+		buffer_c = buffer_acquire(buffer);
+
+		/* if cd->config.out_channels_count are equal to 0
+		 * (it can vary), we set params->channels to sink buffer
+		 * channels, which were previously set in
+		 * pipeline_comp_hw_params()
+		 */
+		out_channels = cd->config.out_channels_count ?
+			cd->config.out_channels_count : buffer_c->stream.channels;
+		params->channels = out_channels;
+	} else {
+		/* fetch source buffer for capture */
+		buffer = list_first_item(&dev->bsource_list, struct comp_buffer,
+					 sink_list);
+		out_channels = cd->config.out_channels_count;
+
+		buffer_c = buffer_acquire(buffer);
+
+		/* if cd->config.in_channels_count are equal to 0
+		 * (it can vary), we set params->channels to source buffer
+		 * channels, which were previously set in
+		 * pipeline_comp_hw_params()
+		 */
+		in_channels = cd->config.in_channels_count ?
+			cd->config.in_channels_count : buffer_c->stream.channels;
+		params->channels = in_channels;
+	}
+
+	/* Set buffer params */
+	buffer_set_params(buffer_c, params, BUFFER_UPDATE_FORCE);
+
+	buffer_release(buffer_c);
+
+	/* set component period frames */
+	sink_c = buffer_acquire(sinkb);
+	component_set_nearest_period_frames(dev, sink_c->stream.rate);
```

Comment:
this should be done even if `in_channels` or `out_channels` turn out to be invalid below?

Reply:
Moved channels count validation to the top of the function.